### ISSUE TITLE
Fallback resolver & Latency 

### ISSUE THREAD

**Issue by CHEF-KOCH** [jedisct1/dnscrypt-proxy#115]
The greatest weakness of DNCrypt is for most people the latency, now we have a fallback resolver mechanism but the main issue isn't fixed. Whenever a resolver/server is slow it tries now to switch to another server, however, this is theoretically a good idea but in fact, it ends up with lots of roundtrips and you might see already in your Browser a DNS error until another resolver was found.

I gave DNSCrypt-Proxy (as a test for an article) 5 different people with different skills and 4 made a complaint about the same thing, the latency. So how we are going to resolve this issue? My idea would be that there could be a check/configuration option to ensure the resolver is not slower than let's say 60 ms which would also already sort out lots of servers on the list.

The problem is that latency is really an important thing because most people which simply want to browse or play online games they care more about this part than the 'security' aspect, in most cases from what I know it ends up that people going to disable it for this time but a workaround to solve this would be a better solution.

I would be thankful for ideas and a discussion ground because 'then disable it' is not really what I like to hear because that's still in my eyes a no-go. I do not really expect low-latency but at least no errors and or disconnects would be fine and I think lots of people could life with that.

**Comment by jedisct1** [comment]
The main issue with DNS, especially over UDP, is that when you don't get a response, there are no reliable way to know what happened. Was the response lost? Is the server slow? Did it ignore this specific query, but is otherwise fine with other query types or queries for other domains?

The only way to know if the server is actually slow or dead is to perform multiple attempts to gradually converge towards an educated guess.

After the proxy is started, there is a warmup period, where all the resolvers are given an equal probability to be chosen as the current best resolver. That warmup period can be shortened. On hosts that are not very chatty, the current value can be too short, but on hosts sending a lot of queries after a network change (such as macOS), it's too long.

That can be a knob, as well as the maximum latency after which we ignore a server. But these knobs would be hard to tune correctly.

Something I'd like to implement is serving cold responses, possibly from a persistent cache. That is, after a timeout, if we sent the same request previously, and even if more time passed than the TTL for this response, we return the response instead of an error. As obvious as it sounds, and in spite of this becoming a common practice, there might be some patents around it, though :(

**Comment by mirsev** [comment]
When I was using dnscrypt-proxy v1, I started 4 processes configured to use different servers and different local addresses. Below there was a dnsmasq which used these 4 local addresses and an option "all servers" in its config file. So, for every DNS query dnsmasq resent these to 4 different DNScrypt-proxies and used first received answer. This configuration was very fast with really low latency. Of course, using 4 different servers means unnecessary network and servers load but I think that this was compensated by using dnsmasq cache.

Unfortunately I still haven't time to study the code for load balancing features of the dnscrypt-proxy v2 (and I don't have much knowledge of Go), so I can only suggest a similar idea: what if dnscrypt-proxy asks not just one but several servers for each query, returns first answer to the consumer and at the same time gathers statistics (server reachability, latency, etc.) for future use to find a pool of most accessible servers.

Is there any docs regarding load balancing, requesting strategy, etc.? It would by quite interesting to read.

**Comment by jedisct1** [comment]
Broadcast every single query to all (currently 67) servers? Please, no, you're not the only user of these servers.
This also means that each query has to be encrypted and decrypted for every single server. That would be very bandwidth and CPU intensive, even client side.

**Comment by mirsev** [comment]
Of course, not to all servers, just a few. And, perhaps this might be needed only for first one hundred requests or so. After that you will have some statistics which shows the best servers for your location and you can use one or two. This might really speed things up. And, since you have cache, I don't think that load will be too high.

Just a note: when I switched from v1 to v2, the DNS requests sometimes take tens of seconds, and sometimes the address was not found though if I check "dig host @8.8.8.8" it returns valid address.

**Comment by mirsev** [comment]
May I show an example when relying on selected by dnscrypt-proxy v2 server is not working?

`mycomp$ host linuxjournal.cmail19.com
Host linuxjournal.cmail19.com not found: 3(NXDOMAIN)
mycomp$ dig linuxjournal.cmail19.com 

; <<>> DiG 9.10.3-P4-Ubuntu <<>> linuxjournal.cmail19.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 22060
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1280
;; QUESTION SECTION:
;linuxjournal.cmail19.com.      IN      A

;; Query time: 0 msec
;; SERVER: 127.0.1.1#53(127.0.1.1)
;; WHEN: Fri Feb 16 21:48:10 MSK 2018
;; MSG SIZE  rcvd: 53

mycomp$ dig linuxjournal.cmail19.com @8.8.8.8

; <<>> DiG 9.10.3-P4-Ubuntu <<>> linuxjournal.cmail19.com @8.8.8.8
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 35887
;; flags: qr rd ra; QUERY: 1, ANSWER: 4, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;linuxjournal.cmail19.com.      IN      A

;; ANSWER SECTION:
linuxjournal.cmail19.com. 299   IN      CNAME   e.csd.link.
e.csd.link.             200     IN      CNAME   edge-pr-alb-r9paf8l6h0p7-150314959.eu-central-1.elb.amazonaws.com.
edge-pr-alb-r9paf8l6h0p7-150314959.eu-central-1.elb.amazonaws.com. 59 IN A 52.57.47.103
edge-pr-alb-r9paf8l6h0p7-150314959.eu-central-1.elb.amazonaws.com. 59 IN A 52.57.119.11

;; Query time: 71 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Fri Feb 16 21:48:25 MSK 2018
;; MSG SIZE  rcvd: 185
`
In this situation, the only possible resolver was dnscrypt-proxy v2. So, does it mean that using only one server selected by dnscrypt-proxy produces the problem? Or, other solution can be suggested for negative answers? What about asking two or more servers in parallel?

**Comment by jedisct1** [comment]
Looks like dnscrypt-proxy just relayed what it received. 

It received `NXDOMAIN`. So it relayed `NXDOMAIN`. There are no timeouts or `SERVFAIL` here.

Asking the same question to multiple servers is probably the worse strategy we can use.

Try different resolvers maybe?

**Comment by jedisct1** [comment]
You can check what resolvers is currently being used with `dnscrypt-proxy -resolve linuxjournal.cmail19.com`

**Comment by mirsev** [comment]
Sure, it relayed what it received,  NXDOMAIN, but why? Host exists, according to google nameserver.

**Comment by mirsev** [comment]
Why Cisco? I'm writing about the setup on my Linux computer.

**Comment by jedisct1** [comment]
Do you know what resolvers are returning `NXDOMAIN` for this query?

**Comment by mirsev** [comment]
How can I ask a dnscrypt-proxy running in background about any particular query?

**Comment by mirsev** [comment]
BTW, you have a list of all available DNS-crypt resolvers... Why don't you ask all of them about this query?

**Comment by jedisct1** [comment]
`dnscrypt-proxy -resolve linuxjournal.cmail19.com`

**Comment by mirsev** [comment]
$ /opt/dnscrypt-proxy/dnscrypt-proxy -resolve linuxjournal.cmail19.com
Resolving [linuxjournal.cmail19.com]

Domain exists:  probably not, or blocked by the proxy
Canonical name: -
IP addresses:   -
TXT records:    -
Resolver IP:    185.140.114.50

**Comment by jedisct1** [comment]
```sh
$ drill linuxjournal.cmail19.com @185.140.114.51

;; ->>HEADER<<- opcode: QUERY, rcode: NXDOMAIN, id: 47678
```

Not even using DNSCrypt at all here. Looks like this resolver (CS Spain) is buggy or has been blacklisted by linuxjournal servers.

Not sure why. I heard a while back that Cryptostorm was filtering some domain names. This might actually be the case, and the exact criteria are unclear.

Can you try to confirm that with them? If they are filtering things, the `filtering` flag must be added to their properties.

**Comment by jedisct1** [comment]
Looks like this is what they do: https://cryptostorm.org/viewtopic.php?f=46&t=8981

The intent is to filter ads and trackers but there might be false positives.

So I guess we need to add the filter flag to all of them :/

**Comment by jedisct1** [comment]
Done.

**Comment by Biggizen** [comment]
A bit offtopic: Could you please explain, what the different load-balancing strategies in the *.toml do? A short non-technical answer would be enough, than you. ;-)

**Comment by jedisct1** [comment]
This should be in the documentation :)

Ok, this thread has been hijacked and doesn't really make sense any more. Gonna open a new ticket dedicated to load balancing instead.
